### ISSUE TITLE
Unable to create endpoint: Cilium API client timeout exceeded

### ISSUE THREAD

**Issue by FranAguiar** [cilium/cilium#32399]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

Since a few weeks ago, some pods in my GKE cluster have been getting stuck in the PodCreating state. When I run a kubectl pod describe, I get this error:

```
 Warning  FailedCreatePodSandBox  4m45s (x138 over 4h)  kubelet  (combined from similar events): Failed to create pod sandbox: rpc error: code = Unknown desc = failed to setup network for sandbox "6eec890c6d2dbfefc3fa6ab1bf8db4f81ccb0c2f53ad757fb8f573a2bf9eca68": plugin type="cilium-cni" failed (add): unable to create endpoint: Cilium API client timeout exceeded
```

And in the logs of the cilium-agent container I found this:

```
{"containerID":"","datapathPolicyRevision":0,"desiredPolicyRevision":0,"endpointID":2588,"error":"timeout while waiting for initial endpoint generation to complete: context canceled","ipv4":"","ipv6":"","k8sPodName":"/","level":"warning","msg":"Creation of endpoint failed","subsys":"daemon"}
```
```
{"containerID":"","datapathPolicyRevision":0,"desiredPolicyRevision":0,"endpointID":795,"error":"unable to resolve identity: failed to assign a global identity for lables: k8s:app.kubernetes.io/component=prometheus,k8s:app.kubernetes.io/instance=monitoring-kube-prometheus-prometheus,k8s:app.kubernetes.io/managed-by=prometheus-operator,k8s:app.kubernetes.io/name=prometheus,k8s:app.kubernetes.io/version=2.48.1,k8s:io.cilium.k8s.namespace.labels.kubernetes.io/metadata.name=monitoring,k8s:io.cilium.k8s.policy.cluster=default,k8s:io.cilium.k8s.policy.serviceaccount=monitoring-kube-prometheus-prometheus,k8s:io.kubernetes.pod.namespace=monitoring,k8s:operator.prometheus.io/name=monitoring-kube-prometheus-prometheus,k8s:operator.prometheus.io/shard=0,k8s:prometheus=monitoring-kube-prometheus-prometheus","identityLabels":{"app.kubernetes.io/component":{"key":"app.kubernetes.io/component","value":"prometheus","source":"k8s"},"app.kubernetes.io/instance":{"key":"app.kubernetes.io/instance","value":"monitoring-kube-prometheus-prometheus","source":"k8s"},"app.kubernetes.io/managed-by":{"key":"app.kubernetes.io/managed-by","value":"prometheus-operator","source":"k8s"},"app.kubernetes.io/name":{"key":"app.kubernetes.io/name","value":"prometheus","source":"k8s"},"app.kubernetes.io/version":{"key":"app.kubernetes.io/version","value":"2.48.1","source":"k8s"},"io.cilium.k8s.namespace.labels.kubernetes.io/metadata.name":{"key":"io.cilium.k8s.namespace.labels.kubernetes.io/metadata.name","value":"monitoring","source":"k8s"},"io.cilium.k8s.policy.cluster":{"key":"io.cilium.k8s.policy.cluster","value":"default","source":"k8s"},"io.cilium.k8s.policy.serviceaccount":{"key":"io.cilium.k8s.policy.serviceaccount","value":"monitoring-kube-prometheus-prometheus","source":"k8s"},"io.kubernetes.pod.namespace":{"key":"io.kubernetes.pod.namespace","value":"monitoring","source":"k8s"},"operator.prometheus.io/name":{"key":"operator.prometheus.io/name","value":"monitoring-kube-prometheus-prometheus","source":"k8s"},"operator.prometheus.io/shard":{"key":"operator.prometheus.io/shard","value":"0","source":"k8s"},"prometheus":{"key":"prometheus","value":"monitoring-kube-prometheus-prometheus","source":"k8s"}},"ipv4":"","ipv6":"","k8sPodName":"/","level":"warning","msg":"Error changing endpoint identity","subsys":"endpoint"}
```

If I manually delete the pod it start without any issue.

### Cilium Version

Client: 1.13.12 38d04fa903 2024-04-05T00:06:43+00:00 go version go1.21.8 linux/amd64
Daemon: 1.13.12 38d04fa903 2024-04-05T00:06:43+00:00 go version go1.21.8 linux/amd64

KVStore:                 Ok   Disabled
Kubernetes:              Ok   1.29 (v1.29.4-gke.1043000) [linux/amd64]
Kubernetes APIs:         ["cilium/v2::CiliumLocalRedirectPolicy", "cilium/v2::CiliumNode", "cilium/v2alpha1::CiliumEndpointSlice", "core/v1::Namespace", "core/v1::Node", "core/v1::Pods", "core/v1::Service", "discovery/v1::EndpointSlice", "networking.k8s.io/v1::NetworkPolicy"]
KubeProxyReplacement:    Strict   [eth0 10.3.2.219 (Direct Routing)]
Host firewall:           Disabled
CNI Chaining:            generic-veth
CNI Config file:         CNI configuration file management disabled
Cilium:                  Ok   1.13.12 (v1.13.12-38d04fa903)
NodeMonitor:             Listening for events on 8 CPUs with 64x4096 of shared memory
IPAM:                    IPv4: 0/62 allocated from 10.3.48.192/26,
IPv6 BIG TCP:            Disabled
BandwidthManager:        EDT with BPF [CUBIC] [eth0]
Host Routing:            Legacy
Masquerading:            Disabled
Controller Status:       77/77 healthy
Proxy Status:            OK, ip 169.254.4.6, 0 redirects active on ports 10000-20000
Global Identity Range:   min 256, max 65535
Hubble:                  Ok   Current/Max Flows: 63/63 (100.00%), Flows/s: 27.13   Metrics: Ok
Encryption:              Disabled
Cluster health:             Probe disabled

### Kernel Version

6.1.75+ #1 SMP PREEMPT_DYNAMIC Sat Mar 30 14:38:17 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux

### Kubernetes Version

Client Version: v1.28.2
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
Server Version: v1.29.4-gke.1043000

### Anything else?

The cluster is in GKE with Dataplane V2 enabled. I don't have control over the cilium agents, it's managed by google. This is only happening on my cluster in the rapid channel, so I'm not sure if it's a bug or some incompatibility between the cilium version and the services affected. 

I can't replicate the error and have no clue about the root cause. Any help will be appreciated, thanks in advance!!

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by Sh4d1** [comment]
Hey, I literally got the same issue, I also opened a support case on GCP, see if it helps, I'll update here if I have some news!

**Comment by FranAguiar** [comment]
Oh, that's great. Please keep me posted. 
I also noticed that this only happens to stuff deployed with Helm (maybe it's just a coincidence, but who knows). Is that the case for you too?

**Comment by Sh4d1** [comment]
Will do! Nope it's deployed with pulumi for me (through native k8s objects).
Do you have a lot of pod changes ? (Like deploys, HPA that are often scaling, ...)

**Comment by FranAguiar** [comment]
No, not in particular. Especially for RabbitMQ, the service that is suffering the most from this issue.

**Comment by r0bj** [comment]
We encountered a similar issue after upgrading our GKE cluster from 1.29.1-gke.1589000 to 1.29.4-gke.1447000. In our case, downgrading the cluster helped, and we were able to fix the issue. @Sh4d1, any luck with GCP support?

**Comment by Sh4d1** [comment]
Interesting! I'm on 1.29.3. And no luck with the support yet (I've linked them the issue as well).

**Comment by harispraba** [comment]
in my case, this happens after node upgrade since we use rapid channel. but not always, sometimes pod recreated successfully.

also, for now this happens only to statefulsets.

**Comment by FranAguiar** [comment]
@harispraba 
Now that you mention it, the services that are experiencing this error in my cluster are also statefulsets. Rabbitmq, thanos-storage, and prometheus.

@Sh4d1 
Hi, any news from GCP support?
Is the ticket public or is it private only with you?

**Comment by Sh4d1** [comment]
Hum I think it's only happened to statefull sets on my end as well! 

@FranAguiar it's private, and no luck yet (they asked for the whole logs from cilium but I don't have it anymore, so waiting for next occurrence to catch it)

**Comment by squeed** [comment]
Assigning to @christarazi, who recently worked on endpoint regeneration and statefulset updates.

**Comment by voltagebots** [comment]
recently faced this as well. In my case, it turns out that Cilium fails to create an endpoint due to the number of character which should be no more than 63. 
Looking at the Cilium agent logs on the node; we see

`[CiliumIdentity.cilium.io](http://ciliumidentity.cilium.io/) \"47595\" is invalid: metadata.labels: Invalid value: \"xxxxxxx": must be no more than 63 characters","key":{"LabelArray":`

renaming the chart name to a fewer characters fixed it.

**Comment by christarazi** [comment]
Can you try v1.15.5? It contains https://github.com/cilium/cilium/pull/31605 which might resolve this problem.

**Comment by FranAguiar** [comment]
Hello, I just update my GKE cluster to latest version
```
Server Version: v1.30.0-gke.1457000
```

And it comes with new cilium version
```
Client: 1.14.7 47ecffbb57 2024-04-25T17:12:33-07:00 go version go1.21.8 linux/amd64
Daemon: 1.14.7 47ecffbb57 2024-04-25T17:12:33-07:00 go version go1.21.8 linux/amd64
```

I hope that solve the issue

**Comment by JJGadgets** [comment]
I have the issue described in OP on 1.15.5 on my _homelab_ baremetal Talos cluster, but it seems to only happen on one node. All pods on that node fails to schedule and the only useful logs are the exact ones reported in OP.

Killing the Cilium pod on that node fixes it for a while until it happens again, and sometimes it happens right from the start of that Cilium pod's lifetime and that pod needds to be killed too. `cilium-dbg status --verbose` shows that this node's endpoints are unreachable, and the other 2 nodes are fine.

Versions:
Talos: 1.6.4
Kubernetes: 1.29.2
Cilium: 1.15.5

Cilium Helm values (these 2 files get merged by Flux Helm controller, and the hr.yaml will override the config/biohazard/helm-values.yaml if there are conflicting values): https://github.com/JJGadgets/Biohazard/blob/7004140fc1be893e1e35dac1d43148af749eb8da/kube/deploy/core/_networking/cilium/app/hr.yaml https://github.com/JJGadgets/Biohazard/blob/7004140fc1be893e1e35dac1d43148af749eb8da/kube/deploy/core/_networking/cilium/app/config/biohazard/helm-values.yaml

**Comment by christarazi** [comment]
@JJGadgets Are the workloads statefulsets?

**Comment by JJGadgets** [comment]
@christarazi Nope, everything from deployments, daemonsets, jobs, to KubeVirt VMs (which is a custom controller AFAIK).

**Comment by christarazi** [comment]
Ok, that sounds like a separate issue from this thread. It seems that the initial report is for statefulsets. I would encourage you to file a new issue with a sysdump of when the issue occurred.

**Comment by JJGadgets** [comment]
@christarazi will create the separate issue when I encounter the issue again, for now the node and its Cilium pod is happy.

**Comment by FranAguiar** [comment]
> Hello, I just update my GKE cluster to latest version
> 
> ```
> Server Version: v1.30.0-gke.1457000
> ```
> 
> And it comes with new cilium version
> 
> ```
> Client: 1.14.7 47ecffbb57 2024-04-25T17:12:33-07:00 go version go1.21.8 linux/amd64
> Daemon: 1.14.7 47ecffbb57 2024-04-25T17:12:33-07:00 go version go1.21.8 linux/amd64
> ```
> 
> I hope that solve the issue

Just happen again, this is the log from the cilium container
[Explore-logs-2024-05-17 09_20_52.txt](https://github.com/cilium/cilium/files/15346054/Explore-logs-2024-05-17.09_20_52.txt)

**Comment by christarazi** [comment]
Any version below 1.15.5 will not have the statefulset fix, so please try upgrading to that.

**Comment by r0bj** [comment]
In my case, not only were stateful sets affected, but deployments were as well.

**Comment by christarazi** [comment]
@r0bj That sounds like a separate issue as mentioned in https://github.com/cilium/cilium/issues/32399#issuecomment-2115851239

**Comment by liquidiert** [comment]
Hey! I have a similiar problem with some of my pods in a AWS cluster. This is the error message:
```text
kubelet  (combined from similar events): Failed to create pod sandbox: rpc error: code = Unknown desc = failed to setup network for sandbox "f40f8a986e4c53a727336b51fb698fee152f06b3357da0079a5ed204ed7d22a0": plugin type="cilium-cni" failed (add): unable to create endpoint: Cilium API client timeout exceeded
```
Cilium verison is `1.16.0-dev`. Anyone encountered the same issue with that verison as well?

**Comment by michalschott** [comment]
> Hey! I have a similiar problem with some of my pods in a AWS cluster. This is the error message:
> 
> ```
> kubelet  (combined from similar events): Failed to create pod sandbox: rpc error: code = Unknown desc = failed to setup network for sandbox "f40f8a986e4c53a727336b51fb698fee152f06b3357da0079a5ed204ed7d22a0": plugin type="cilium-cni" failed (add): unable to create endpoint: Cilium API client timeout exceeded
> ```
> 
> Cilium verison is `1.16.0-dev`. Anyone encountered the same issue with that verison as well?

Yes on 1.15.6.

We had to add alert for pods being in init phase for longer than X and the only way to fix this is to manually delete pod to reschedule it on different node (cordon faulty node as well).

**Comment by robpearce-flux** [comment]
seeing this error on eks 1.30 in aws, cilium 1.15.5. Not sure how to reproduce yet, i have 8 clusters, it's happening on several of them, in one case with the ca-injector pod of the cert-manager service
```  Warning  FailedCreatePodSandBox  63s (x14 over 23m)  kubelet            (combined from similar events): Failed to create pod sandbox: rpc error: code = Unknown desc = failed to setup network for sandbox "fcbcd10d03a65468af8d17915321cd806aeabcb1645b45258e7d044d608b5555": plugin type="cilium-cni" failed (add): unable to create endpoint: Cilium API client timeout exceeded
```

**Comment by hamza-cskn** [comment]
same error on rke2 v1.30.5 (on-premise), cilium v1.16.2. I had successfully run following config last week but today I can't. I changed tons of thing. I am not sure how to reproduce.


(using deployments instead of pods, did not fix)
```yaml
apiVersion: v1
kind: Pod
metadata:
  name: requester-pod
spec:
  containers:
  - name: requester
    image: curlimages/curl:7.85.0
    command: ["/bin/sh", "-c", "while true; do curl http://nginx-service/; sleep 2; done"]
```
```logs
3m30s       Normal    Scheduled                Pod/requester-pod   Successfully assigned default/requester-pod to d4rke-wrk02
2m          Warning   FailedCreatePodSandBox   Pod/requester-pod   Failed to create pod sandbox: rpc error: code = Unknown desc = failed to setup network for sandbox "df248c596c34bf0ce6c41883a8c5732ca7d2524b3e3a69d8c65fd835b6c8fd13": plugin type="cilium-cni" failed (add): unable to create endpoint: Cilium API client timeout exceeded
16s         Warning   FailedCreatePodSandBox   Pod/requester-pod   Failed to create pod sandbox: rpc error: code = Unknown desc = failed to setup network for sandbox "b4583144d09ee7496c61e0cded80f39f20b7e9a8c34a4845a23a866e931f06b9": plugin type="cilium-cni" failed (add): unable to create endpoint: Cilium API client timeout exceeded
```

**Comment by tschlaepfer** [comment]
I get the same issue from time to time on EKS 1.30 with Cilium v1.62.2. I also cannot reproduce it consistently but in my case it is always with the ingress-nginx controller.

The ingress-nginx-controller pods return the following events:
```
Failed to create pod sandbox: rpc error: code = Unknown desc = failed to setup network for sandbox "b065e48d2e7f55b1a97e7d13a7a8f2aa49de02851853f18bfa2f86bc92fabb66": plugin type="cilium-cni" failed (add): unable to create endpoint: Cilium API client timeout exceeded
```

If I check the logs of the cilium agent running on the same node I see the following error message:
```
level=error msg="endpoint regeneration failed" ciliumEndpointName=ingress-nginx/ingress-nginx-controller-8475fd878-2nrtm containerID=ecf5563682 containerInterface= datapathPolicyRevision=0 desiredPolicyRevision=0 endpointID=698 error="unable to regenerate policy for '698': secrets \"vlinsert-server-tls\" not found" identity=17797 ipv4=10.100.15.4 ipv6="fe80::6cbe:74ff:fe54:8ada" k8sPodName=ingress-nginx/ingress-nginx-controller-8475fd878-2nrtm subsys=endpoint

level=warning msg="policy: Error getting terminating TLS Context." error="secrets \"vlinsert-server-tls\" not found" subsys=policy
time="2024-11-12T11:05:21Z" level=warning msg="Failed to update policy" ciliumEndpointName=ingress-nginx/ingress-nginx-controller-8475fd878-2nrtm containerID=ecf5563682 containerInterface= datapathPolicyRevision=0 desiredPolicyRevision=0 endpointID=698 error="secrets \"vlinsert-server-tls\" not found" identity=17797 ipv4=10.100.15.4 ipv6="fe80::6cbe:74ff:fe54:8ada" k8sPodName=ingress-nginx/ingress-nginx-controller-8475fd878-2nrtm subsys=endpoint

level=info msg="generating BPF for endpoint failed, keeping stale directory" ciliumEndpointName=ingress-nginx/ingress-nginx-controller-8475fd878-2nrtm containerID=ecf5563682 containerInterface= datapathPolicyRevision=0 desiredPolicyRevision=0 endpointID=698 error="unable to regenerate policy for '698': secrets \"vlinsert-server-tls\" not found" file-path=698_next_fail identity=17797 ipv4=10.100.15.4 ipv6="fe80::6cbe:74ff:fe54:8ada" k8sPodName=ingress-nginx/ingress-nginx-controller-8475fd878-2nrtm subsys=endpoint

level=warning msg="Regeneration of endpoint failed" bpfCompilation=0s bpfLoadProg=0s bpfWaitForELF=0s ciliumEndpointName=ingress-nginx/ingress-nginx-controller-8475fd878-2nrtm containerID=ecf5563682 containerInterface= datapathPolicyRevision=0 desiredPolicyRevision=0 endpointID=698 error="unable to regenerate policy for '698': secrets \"vlinsert-server-tls\" not found" identity=17797 ipv4=10.100.15.4 ipv6="fe80::6cbe:74ff:fe54:8ada" k8sPodName=ingress-nginx/ingress-nginx-controller-8475fd878-2nrtm mapSync=0s policyCalculation=13.35337ms prepareBuild="51.553µs" proxyConfiguration=0s proxyPolicyCalculation=0s proxyWaitForAck=0s reason="retrying regeneration" subsys=endpoint total=13.604785ms waitingForCTClean=0s waitingForLock="1.517µs" waitingForPolicyRepository=480ns

level=error msg="endpoint regeneration failed" ciliumEndpointName=ingress-nginx/ingress-nginx-controller-8475fd878-2nrtm containerID=ecf5563682 containerInterface= datapathPolicyRevision=0 desiredPolicyRevision=0 endpointID=698 error="unable to regenerate policy for '698': secrets \"vlinsert-server-tls\" not found" identity=17797 ipv4=10.100.15.4 ipv6="fe80::6cbe:74ff:fe54:8ada" k8sPodName=ingress-nginx/ingress-nginx-controller-8475fd878-2nrtm subsys=endpoint
```

I found that, in my case, the issue seems somehow related to the network policies from Cilium. Deleting all the Cilium network policies allowed the ingress-nginx pods to be scheduled. Afterwards reapplying the Cilium network policies does not cause any issues. I only experienced this issue during the initial deployment of my Kubernetes platform.

Hope this may be helpful to others or debug the issue in Cilium.

**Comment by gadiener** [comment]
The issue recurred today when the Cilium agent entered a crash loop due to an out-of-memory (OOM) condition. After increasing the resources allocated to the Cilium agent, the Datadog pods became stuck in the initialization phase, and the log message unable to regenerate policy because PolicyMap synchronization failed: syncPolicyMap failed appeared across all Cilium agents, including those on newly created nodes.

The problem was temporarily resolved by deleting all CiliumNetworkPolicy, but it come back with a new log
```
{"bpfMapKey":{"Prefixlen":40,"Identity":7,"TrafficDirection":1,"Nexthdr":0,"DestPortNetwork":0},"ciliumEndpointName":"redacted","containerID":"139306ca4d","containerInterface":"","datapathPolicyRevision":0,"desiredPolicyRevision":55,"endpointID":326,"error":"update map cilium_policy_00326: update: no space left on device","identity":"35892","ipv4":"redacted","ipv6":"","k8sPodName":"redacted","level":"error","msg":"Failed to add PolicyMap key", "port":0,"subsys":"endpoint"}
```

An interesting detail, an allow-all `CiliumNetworkPolicy` is present in a namespace: 
```
spec:
  egress:
  - toFQDNs:
    - matchPattern: '*'
  - toEndpoints:
    - matchLabels:
        k8s:io.kubernetes.pod.namespace: kube-system
        k8s:k8s-app: kube-dns
    toPorts:
    - ports:
      - port: "53"
        protocol: UDP
      - port: "53"
        protocol: TCP
      rules:
        dns:
        - matchPattern: '*'
  endpointSelector:
    matchLabels: {}
```
Deleting that seems to help

**Comment by christarazi** [comment]
`syncPolicyMap failed` occurs if Cilium fails to add/delete entries from the policymap. ` no space left on device` suggests that the policymaps are full which sounds like the root cause. I would suggest following https://docs.cilium.io/en/latest/operations/troubleshooting/#policymap-pressure-and-overflow to see if you can make further progress on understanding why the policymaps are filling up.

**Comment by gadiener** [comment]
Yes @christarazi that was indeed the case, running Airflow we had many pods with high cardinality labels containing random IDs or timestamps. Fixed by limiting the [identity-relevant labels](https://docs.cilium.io/en/latest/operations/performance/scalability/identity-relevant-labels/#identity-relevant-labels) in the Cilium configuration.
